### ISSUE TITLE
[BUG] - AfterAll And AfterEach Not Working Properly

### ISSUE THREAD

**Issue by danrealednb** [microsoft/playwright#27457]
### System info
- Playwright Version: [v1.38.1]
- Operating System: [macOS 13.5.1]
- Browser: [Chromium]
- Other info:

### Source code

- [x] I provided exact source code that allows reproducing the issue locally.

**Link to the GitHub repository with the repro**

[https://github.com/your_profile/playwright_issue_title]

or

**Config file**

```js
import { defineConfig, devices } from "@playwright/test";
import "dotenv/config";
require("dotenv").config();
/**
 * Read environment variables from file.
 * https://github.com/motdotla/dotenv
 */
// require('dotenv').config();

/**
 * See https://playwright.dev/docs/test-configuration.
 */

export const getBaseUrl = () => {
  const test_env = process.env.TEST_ENV || "local";
  switch (test_env) {
    case "local":
      return "";
    case "dev":
      return "";
    case "qa":
      return "https://localhost:3000";
    case "staging":
      return "https://localhost:3000";
    case "prod":
      return "https://localhost:3000";
    default:
      throw new Error(`${test_env} is not a valid test environment`);
  }
};

export default defineConfig({
  timeout: 5 * 60 * 1000, // 5 minutes
  globalTimeout: undefined,
  testDir: "./tests",
  // testIgnore: "**/Tools/*spec.ts", // uncomment this when wanting to run Tools tests
  /* Run tests in files in parallel */
  fullyParallel: false,
  /* Fail the build on CI if you accidentally left test.only in the source code. */
  forbidOnly: !!process.env.CI,
  /* Retry on CI only */
  retries: process.env.CI ? 2 : 0,
  /* Opt out of parallel tests on CI. */
  // set undefined to run parallel with max workeres
  workers: process.env.CI ? 1 : 1,
  /* Reporter to use. See https://playwright.dev/docs/test-reporters */
  reporter: [
    ["html", { open: "never" }],
    ["list"],
    ["json", { outputFile: "test-results.json" }],
  ],
  /* Shared settings for all the projects below. See https://playwright.dev/docs/api/class-testoptions. */
  use: {
    /* Base URL to use in actions like `await page.goto('/')`. */
    // baseURL: 'http://127.0.0.1:3000',

    /* Collect trace when retrying the failed test. See https://playwright.dev/docs/trace-viewer */
    trace: "retain-on-failure",
    actionTimeout: 10000,
    navigationTimeout: 60000,
    screenshot: "only-on-failure",
  },
  reportSlowTests: { max: 10, threshold: 60000 },

  /* Configure projects for major browsers */
  projects: [
    {
      name: "chromium",
      use: {
        ...devices["Desktop Chrome"],
        headless: false,
        baseURL: getBaseUrl(),
      },
    },

    // {
    //   name: "firefox",
    //   use: {
    //     ...devices["Desktop Firefox"],
    //     headless: false,
    //     baseURL: getBaseUrl(),
    //   },
    // },

    // {
    //   name: "webkit",
    //   use: {
    //     ...devices["Desktop Safari"],
    //     headless: false,
    //     baseURL: getBaseUrl(),
    //   },
    // },

    /* Test against mobile viewports. */
    // {
    //   name: 'Mobile Chrome',
    //   use: { ...devices['Pixel 5'] },
    // },
    // {
    //   name: 'Mobile Safari',
    //   use: { ...devices['iPhone 12'] },
    // },

    /* Test against branded browsers. */
    // {
    //   name: 'Microsoft Edge',
    //   use: { ...devices['Desktop Edge'], channel: 'msedge' },
    // },
    // {
    //   name: 'Google Chrome',
    //   use: { ...devices['Desktop Chrome'], channel: 'chrome' },
    // },
  ],

  /* Run your local dev server before starting the tests */
  // webServer: {
  //   command: 'npm run start',
  //   url: 'http://127.0.0.1:3000',
  //   reuseExistingServer: !process.env.CI,
  // },
});

```

**Test file (self-contained)**

```js
import { test, type Page, expect } from "@playwright/test";
import { queryTable } from "@pom/db";
const fs = require("fs");

// test.describe.configure({ mode: "parallel" });
test.describe("NAICS Codes Database Mapping", () => {
  let code: string;
  const failedCodes = [];
  test.beforeEach(async () => {
    code = "none";
  });
  test.afterEach(async ({}, testInfo) => {
    if (testInfo.status !== "passed") {
      console.log(`Code: ${code} failed`);
      failedCodes.push(code);
      console.log(failedCodes);
    }
  });
  test.afterAll(async () => {
    console.log("Done with test");
    writer("NAICS_Mapping_Failures.json", failedCodes);
  });
  const mapping_file = require("../../fixtures/naics_codes_2017_to_2022.json");
  for (let i = 0; i < 100; i++) {
    const element = mapping_file[i];
    test(`Test The Database Mapping 2017 Code ${element.code_2017} Id: ${element.id}`, async () => {
      code = element.code_2017;
      const dbResults = await queryTable(
        `select * from table where ncode = '${element.code_2017}'`
      );
      expect(dbResults[0].Naics_2017_Code).toEqual(element.code_2017);

      expect(dbResults[0].Naics_2017_Desc).toContain(element.code_2017);
      expect(dbResults[0].Naics_2017_Desc).toContain(element.description_2017);

      expect(dbResults[0].Naics_2022_Code).toEqual(element.code_2022);

      expect(dbResults[0].Naics_2022_Desc).toContain(element.code_2022);
      expect(dbResults[0].Naics_2022_Desc).toContain(element.description_2022);
    });
  }
});
function writer(fileName: string, data: Array<string>) {
  fs.writeFile(fileName, JSON.stringify(data, null, 2), function (err: any) {
    if (err) {
      return console.log(err);
    }
    console.log(`${fileName} was created successfully!`);
  });
}

```

**Steps**
- `npx playwright test tests/Tools/NAICS_DB_Mapping.spec.ts`
- Check report

```
> TEST_ENV=qa npx playwright test tests/Tools/NAICS_DB_Mapping.spec.ts


Running 100 tests using 1 worker

  ✓  1 …um] › Tools/NAICS_DB_Mapping.spec.ts:26:9 › NAICS Codes Database Mapping › Test The Database Mapping 2017 Code 111110 Id: 1 (380ms)
  ✓  2 …um] › Tools/NAICS_DB_Mapping.spec.ts:26:9 › NAICS Codes Database Mapping › Test The Database Mapping 2017 Code 111120 Id: 2 (385ms)
  ✓  3 …um] › Tools/NAICS_DB_Mapping.spec.ts:26:9 › NAICS Codes Database Mapping › Test The Database Mapping 2017 Code 111130 Id: 3 (426ms)
  ✓  4 …um] › Tools/NAICS_DB_Mapping.spec.ts:26:9 › NAICS Codes Database Mapping › Test The Database Mapping 2017 Code 111140 Id: 4 (453ms)
  ✓  5 …um] › Tools/NAICS_DB_Mapping.spec.ts:26:9 › NAICS Codes Database Mapping › Test The Database Mapping 2017 Code 111150 Id: 5 (386ms)
  ✓  6 …um] › Tools/NAICS_DB_Mapping.spec.ts:26:9 › NAICS Codes Database Mapping › Test The Database Mapping 2017 Code 111160 Id: 6 (375ms)
  ✓  7 …um] › Tools/NAICS_DB_Mapping.spec.ts:26:9 › NAICS Codes Database Mapping › Test The Database Mapping 2017 Code 111191 Id: 7 (431ms)
  ✓  8 …um] › Tools/NAICS_DB_Mapping.spec.ts:26:9 › NAICS Codes Database Mapping › Test The Database Mapping 2017 Code 111199 Id: 8 (408ms)
  ✓  9 …um] › Tools/NAICS_DB_Mapping.spec.ts:26:9 › NAICS Codes Database Mapping › Test The Database Mapping 2017 Code 111211 Id: 9 (456ms)
  ✓  10 …] › Tools/NAICS_DB_Mapping.spec.ts:26:9 › NAICS Codes Database Mapping › Test The Database Mapping 2017 Code 111219 Id: 10 (429ms)
  ✓  11 …] › Tools/NAICS_DB_Mapping.spec.ts:26:9 › NAICS Codes Database Mapping › Test The Database Mapping 2017 Code 111310 Id: 11 (394ms)
  ✓  12 …] › Tools/NAICS_DB_Mapping.spec.ts:26:9 › NAICS Codes Database Mapping › Test The Database Mapping 2017 Code 111320 Id: 12 (401ms)
  ✓  13 …] › Tools/NAICS_DB_Mapping.spec.ts:26:9 › NAICS Codes Database Mapping › Test The Database Mapping 2017 Code 111331 Id: 13 (389ms)
  ✓  14 …] › Tools/NAICS_DB_Mapping.spec.ts:26:9 › NAICS Codes Database Mapping › Test The Database Mapping 2017 Code 111332 Id: 14 (407ms)
  ✓  15 …] › Tools/NAICS_DB_Mapping.spec.ts:26:9 › NAICS Codes Database Mapping › Test The Database Mapping 2017 Code 111333 Id: 15 (434ms)
  ✓  16 …] › Tools/NAICS_DB_Mapping.spec.ts:26:9 › NAICS Codes Database Mapping › Test The Database Mapping 2017 Code 111334 Id: 16 (424ms)
  ✓  17 …] › Tools/NAICS_DB_Mapping.spec.ts:26:9 › NAICS Codes Database Mapping › Test The Database Mapping 2017 Code 111335 Id: 17 (422ms)
  ✓  18 …] › Tools/NAICS_DB_Mapping.spec.ts:26:9 › NAICS Codes Database Mapping › Test The Database Mapping 2017 Code 111336 Id: 18 (422ms)
  ✓  19 …] › Tools/NAICS_DB_Mapping.spec.ts:26:9 › NAICS Codes Database Mapping › Test The Database Mapping 2017 Code 111339 Id: 19 (429ms)
  ✓  20 …] › Tools/NAICS_DB_Mapping.spec.ts:26:9 › NAICS Codes Database Mapping › Test The Database Mapping 2017 Code 111411 Id: 20 (397ms)
  ✓  21 …] › Tools/NAICS_DB_Mapping.spec.ts:26:9 › NAICS Codes Database Mapping › Test The Database Mapping 2017 Code 111419 Id: 21 (397ms)
  ✓  22 …] › Tools/NAICS_DB_Mapping.spec.ts:26:9 › NAICS Codes Database Mapping › Test The Database Mapping 2017 Code 111421 Id: 22 (368ms)
  ✓  23 …] › Tools/NAICS_DB_Mapping.spec.ts:26:9 › NAICS Codes Database Mapping › Test The Database Mapping 2017 Code 111422 Id: 23 (431ms)
  ✓  24 …] › Tools/NAICS_DB_Mapping.spec.ts:26:9 › NAICS Codes Database Mapping › Test The Database Mapping 2017 Code 111910 Id: 24 (375ms)
  ✓  25 …] › Tools/NAICS_DB_Mapping.spec.ts:26:9 › NAICS Codes Database Mapping › Test The Database Mapping 2017 Code 111920 Id: 25 (410ms)
  ✓  26 …] › Tools/NAICS_DB_Mapping.spec.ts:26:9 › NAICS Codes Database Mapping › Test The Database Mapping 2017 Code 111930 Id: 26 (384ms)
  ✓  27 …] › Tools/NAICS_DB_Mapping.spec.ts:26:9 › NAICS Codes Database Mapping › Test The Database Mapping 2017 Code 111940 Id: 27 (373ms)
  ✓  28 …] › Tools/NAICS_DB_Mapping.spec.ts:26:9 › NAICS Codes Database Mapping › Test The Database Mapping 2017 Code 111991 Id: 28 (443ms)
  ✓  29 …] › Tools/NAICS_DB_Mapping.spec.ts:26:9 › NAICS Codes Database Mapping › Test The Database Mapping 2017 Code 111992 Id: 29 (396ms)
  ✓  30 …] › Tools/NAICS_DB_Mapping.spec.ts:26:9 › NAICS Codes Database Mapping › Test The Database Mapping 2017 Code 111998 Id: 30 (397ms)
  ✓  31 …] › Tools/NAICS_DB_Mapping.spec.ts:26:9 › NAICS Codes Database Mapping › Test The Database Mapping 2017 Code 112111 Id: 31 (386ms)
  ✓  32 …] › Tools/NAICS_DB_Mapping.spec.ts:26:9 › NAICS Codes Database Mapping › Test The Database Mapping 2017 Code 112112 Id: 32 (394ms)
  ✓  33 …] › Tools/NAICS_DB_Mapping.spec.ts:26:9 › NAICS Codes Database Mapping › Test The Database Mapping 2017 Code 112120 Id: 33 (385ms)
  ✓  34 …] › Tools/NAICS_DB_Mapping.spec.ts:26:9 › NAICS Codes Database Mapping › Test The Database Mapping 2017 Code 112130 Id: 34 (549ms)
  ✓  35 …] › Tools/NAICS_DB_Mapping.spec.ts:26:9 › NAICS Codes Database Mapping › Test The Database Mapping 2017 Code 112210 Id: 35 (402ms)
  ✓  36 …] › Tools/NAICS_DB_Mapping.spec.ts:26:9 › NAICS Codes Database Mapping › Test The Database Mapping 2017 Code 112310 Id: 36 (377ms)
  ✓  37 …] › Tools/NAICS_DB_Mapping.spec.ts:26:9 › NAICS Codes Database Mapping › Test The Database Mapping 2017 Code 112320 Id: 37 (406ms)
  ✓  38 …] › Tools/NAICS_DB_Mapping.spec.ts:26:9 › NAICS Codes Database Mapping › Test The Database Mapping 2017 Code 112330 Id: 38 (405ms)
  ✓  39 …] › Tools/NAICS_DB_Mapping.spec.ts:26:9 › NAICS Codes Database Mapping › Test The Database Mapping 2017 Code 112340 Id: 39 (386ms)
  ✓  40 …] › Tools/NAICS_DB_Mapping.spec.ts:26:9 › NAICS Codes Database Mapping › Test The Database Mapping 2017 Code 112390 Id: 40 (407ms)
  ✓  41 …] › Tools/NAICS_DB_Mapping.spec.ts:26:9 › NAICS Codes Database Mapping › Test The Database Mapping 2017 Code 112410 Id: 41 (451ms)
  ✓  42 …] › Tools/NAICS_DB_Mapping.spec.ts:26:9 › NAICS Codes Database Mapping › Test The Database Mapping 2017 Code 112420 Id: 42 (422ms)
  ✓  43 …] › Tools/NAICS_DB_Mapping.spec.ts:26:9 › NAICS Codes Database Mapping › Test The Database Mapping 2017 Code 112511 Id: 43 (486ms)
  ✓  44 …] › Tools/NAICS_DB_Mapping.spec.ts:26:9 › NAICS Codes Database Mapping › Test The Database Mapping 2017 Code 112512 Id: 44 (510ms)
  ✓  45 …] › Tools/NAICS_DB_Mapping.spec.ts:26:9 › NAICS Codes Database Mapping › Test The Database Mapping 2017 Code 112519 Id: 45 (411ms)
  ✓  46 …] › Tools/NAICS_DB_Mapping.spec.ts:26:9 › NAICS Codes Database Mapping › Test The Database Mapping 2017 Code 112910 Id: 46 (420ms)
  ✓  47 …] › Tools/NAICS_DB_Mapping.spec.ts:26:9 › NAICS Codes Database Mapping › Test The Database Mapping 2017 Code 112920 Id: 47 (453ms)
  ✓  48 …] › Tools/NAICS_DB_Mapping.spec.ts:26:9 › NAICS Codes Database Mapping › Test The Database Mapping 2017 Code 112930 Id: 48 (795ms)
  ✓  49 …] › Tools/NAICS_DB_Mapping.spec.ts:26:9 › NAICS Codes Database Mapping › Test The Database Mapping 2017 Code 112990 Id: 49 (386ms)
  ✓  50 …] › Tools/NAICS_DB_Mapping.spec.ts:26:9 › NAICS Codes Database Mapping › Test The Database Mapping 2017 Code 113110 Id: 50 (856ms)
  ✓  51 …] › Tools/NAICS_DB_Mapping.spec.ts:26:9 › NAICS Codes Database Mapping › Test The Database Mapping 2017 Code 113210 Id: 51 (383ms)
  ✓  52 …] › Tools/NAICS_DB_Mapping.spec.ts:26:9 › NAICS Codes Database Mapping › Test The Database Mapping 2017 Code 113310 Id: 52 (386ms)
  ✓  53 …] › Tools/NAICS_DB_Mapping.spec.ts:26:9 › NAICS Codes Database Mapping › Test The Database Mapping 2017 Code 114111 Id: 53 (397ms)
  ✓  54 …] › Tools/NAICS_DB_Mapping.spec.ts:26:9 › NAICS Codes Database Mapping › Test The Database Mapping 2017 Code 114112 Id: 54 (390ms)
  ✓  55 …] › Tools/NAICS_DB_Mapping.spec.ts:26:9 › NAICS Codes Database Mapping › Test The Database Mapping 2017 Code 114119 Id: 55 (397ms)
  ✓  56 …] › Tools/NAICS_DB_Mapping.spec.ts:26:9 › NAICS Codes Database Mapping › Test The Database Mapping 2017 Code 114210 Id: 56 (380ms)
  ✓  57 …] › Tools/NAICS_DB_Mapping.spec.ts:26:9 › NAICS Codes Database Mapping › Test The Database Mapping 2017 Code 115111 Id: 57 (391ms)
  ✓  58 …] › Tools/NAICS_DB_Mapping.spec.ts:26:9 › NAICS Codes Database Mapping › Test The Database Mapping 2017 Code 115112 Id: 58 (387ms)
  ✓  59 …] › Tools/NAICS_DB_Mapping.spec.ts:26:9 › NAICS Codes Database Mapping › Test The Database Mapping 2017 Code 115113 Id: 59 (377ms)
  ✓  60 …] › Tools/NAICS_DB_Mapping.spec.ts:26:9 › NAICS Codes Database Mapping › Test The Database Mapping 2017 Code 115114 Id: 60 (386ms)
  ✓  61 …] › Tools/NAICS_DB_Mapping.spec.ts:26:9 › NAICS Codes Database Mapping › Test The Database Mapping 2017 Code 115115 Id: 61 (397ms)
  ✓  62 …] › Tools/NAICS_DB_Mapping.spec.ts:26:9 › NAICS Codes Database Mapping › Test The Database Mapping 2017 Code 115116 Id: 62 (421ms)
  ✓  63 …] › Tools/NAICS_DB_Mapping.spec.ts:26:9 › NAICS Codes Database Mapping › Test The Database Mapping 2017 Code 115210 Id: 63 (438ms)
  ✓  64 …] › Tools/NAICS_DB_Mapping.spec.ts:26:9 › NAICS Codes Database Mapping › Test The Database Mapping 2017 Code 115310 Id: 64 (368ms)
  ✓  65 …] › Tools/NAICS_DB_Mapping.spec.ts:26:9 › NAICS Codes Database Mapping › Test The Database Mapping 2017 Code 211120 Id: 65 (398ms)
  ✓  66 …] › Tools/NAICS_DB_Mapping.spec.ts:26:9 › NAICS Codes Database Mapping › Test The Database Mapping 2017 Code 211130 Id: 66 (421ms)
  ✓  67 …] › Tools/NAICS_DB_Mapping.spec.ts:26:9 › NAICS Codes Database Mapping › Test The Database Mapping 2017 Code 212111 Id: 67 (393ms)
  ✘  68 …] › Tools/NAICS_DB_Mapping.spec.ts:26:9 › NAICS Codes Database Mapping › Test The Database Mapping 2017 Code 212113 Id: 68 (435ms)
Code: 212113 failed
[ '212113' ]
Done with test
NAICS_Mapping_Failures.json was created successfully!
  ✓  69 …] › Tools/NAICS_DB_Mapping.spec.ts:26:9 › NAICS Codes Database Mapping › Test The Database Mapping 2017 Code 212112 Id: 69 (371ms)
  ✘  70 …] › Tools/NAICS_DB_Mapping.spec.ts:26:9 › NAICS Codes Database Mapping › Test The Database Mapping 2017 Code 212113 Id: 70 (461ms)
Code: 212113 failed
[ '212113' ]
Done with test
NAICS_Mapping_Failures.json was created successfully!
  ✓  71 …] › Tools/NAICS_DB_Mapping.spec.ts:26:9 › NAICS Codes Database Mapping › Test The Database Mapping 2017 Code 212210 Id: 71 (528ms)
  ✓  72 …] › Tools/NAICS_DB_Mapping.spec.ts:26:9 › NAICS Codes Database Mapping › Test The Database Mapping 2017 Code 212221 Id: 72 (460ms)
  ✘  73 …] › Tools/NAICS_DB_Mapping.spec.ts:26:9 › NAICS Codes Database Mapping › Test The Database Mapping 2017 Code 212222 Id: 73 (505ms)
Code: 212222 failed
[ '212222' ]
Done with test
NAICS_Mapping_Failures.json was created successfully!
  ✓  74 …] › Tools/NAICS_DB_Mapping.spec.ts:26:9 › NAICS Codes Database Mapping › Test The Database Mapping 2017 Code 212230 Id: 74 (386ms)
  ✓  75 …] › Tools/NAICS_DB_Mapping.spec.ts:26:9 › NAICS Codes Database Mapping › Test The Database Mapping 2017 Code 212299 Id: 75 (398ms)
  ✓  76 …] › Tools/NAICS_DB_Mapping.spec.ts:26:9 › NAICS Codes Database Mapping › Test The Database Mapping 2017 Code 212291 Id: 76 (390ms)
  ✓  77 …] › Tools/NAICS_DB_Mapping.spec.ts:26:9 › NAICS Codes Database Mapping › Test The Database Mapping 2017 Code 212311 Id: 77 (377ms)
```

**Expected**

Expecting the `afterAll` hook to run after all the tests run

**Actual**

The `afterAll` hook is running after each of the tests. Also the `afterEach` seems to be resetting the array and not actually pushing any values to it like I'd expect it to.  Either I'm missing something or there is a bug with when these things are being fired.
Im trying to gather an array of failed test codes so that I can go and debug them but the after each and after all aren't doing what I'd expect them to do.

**Comment by pavelfeldman** [comment]
This looks like working as intended, every time a test fails, we consider environment compromised, entire world is re-created and we start with the next test. You should rely upon custom reporter and test annotations and/or attachments if you want to collect all the failed 'codes' instead.

**Comment by danrealednb** [comment]
I don't believe annotations are the correct approach here since I'd need to know the status of the test before the test runs. How would you approach this from a code perspective?

**Comment by pavelfeldman** [comment]
Same as in what you do, in afterEach you could check the testInfo status and add an annotation or attachment that you would process in the reporter.

**Comment by pavelfeldman** [comment]
Closing as per above, please feel free to open a new issue if this does not cover your use case.
